### PR TITLE
MAINT: clip SLSQP step within bounds

### PR DESCRIPTION
**Pull request by andyfaff**
Addresses a subset of the issues in #11403, see [here](https://github.com/scipy/scipy/issues/11403#issuecomment-717676023) for why this PR is necessary.

### CODE REVIEW

**Comment by andyfaff** [comment]
It would be preferable if SLSQP Fortran code didn't exceed the bounds at all.

**Comment by andyfaff** [comment]
@lukasheinrich, @kratsg can you try out this proposed fix for #11403 and see how it goes on your issues?

**Comment by andyfaff** [comment]
One possibility is that we issue no warning and just clip by default.

**Comment by andyfaff** [comment]
Last commit also applies bounds to the numerical differentiation of the jacobian, as mentioned in @davidpasquale comments in #12949.

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -264,6 +264,20 @@ def hess(x, *args):
...
+    # suggest a move that is outside the limits by 1 or 2 ULP. This
+    # unclean fix makes sure x is strictly within bounds.
+    def eval(x):
+        if (x < bounds[0]).any() or (x > bounds[1]).any():
+            warnings.warn("Values in x were outside bounds during an "
+                          "SLSQP step, clipping to bounds", RuntimeWarning)
+            x = np.clip(x, bounds[0], bounds[1])
+        return func(x)
+    return eval
+
```
Comment (larsoner) [inline scipy/optimize/optimize.py:280]:
```suggestion
        x = _check_clip_x(x, bounds)
        return func(x)
    return eval


def _check_clip_x(x, bounds):
    if (x < bounds[0]).any() or (x > bounds[1]).any():
        warnings.warn("Values in x were outside bounds during an "
                      "SLSQP step, clipping to bounds", RuntimeWarning)
        x = np.clip(x, bounds[0], bounds[1])

```

File: scipy/optimize/slsqp.py [hunk 2]
```diff
@@ -15,12 +15,13 @@
...
 __all__ = ['approx_jacobian', 'fmin_slsqp']
 
+import warnings
 import numpy as np
 from scipy.optimize._slsqp import slsqp
 from numpy import (zeros, array, linalg, append, asfarray, concatenate, finfo,
                    sqrt, vstack, exp, inf, isfinite, atleast_1d)
 from .optimize import (OptimizeResult, _check_unknown_options,
-                       _prepare_scalar_function)
+                       _prepare_scalar_function, clip_x_for_func)
```
Comment (larsoner) [inline scipy/optimize/slsqp.py:24]:
```suggestion
                       _prepare_scalar_function, clip_x_for_func,
                       _check_clip_x)
```

File: scipy/optimize/slsqp.py [hunk 3]
```diff
@@ -277,12 +291,20 @@ def _minimize_slsqp(func, x0, args=(), jac=None, bounds=None,
             # to keep a reference to `fun`, see gh-4240.
             def cjac_factory(fun):
                 def cjac(x, *args):
+                    if (x < new_bounds[0]).any() or (x > new_bounds[1]).any():
+                        warnings.warn("Values in x were outside bounds during"
+                                      " an SLSQP step, clipping to bounds",
+                                      RuntimeWarning)
+                        x = np.clip(x, new_bounds[0], new_bounds[1])
+
```
Comment (larsoner) [inline scipy/optimize/slsqp.py:299]:
```suggestion
                    _check_clip_x(x, bounds)
```

File: scipy/optimize/tests/test_slsqp.py [hunk 4]
```diff
@@ -564,3 +586,22 @@ def test_gh9640(self):
...
+        bounds = Bounds(np.array([0.1]), np.array([1.0]))
+        n_inputs = len(bounds.lb)
+        x0 = np.array(bounds.lb + (bounds.ub - bounds.lb) *
+                      np.random.random(n_inputs))
+
+        def f(x):
+            assert (x >= bounds.lb).all()
+            return np.linalg.norm(x)
+
+        with pytest.warns(RuntimeWarning):
```
Comment (larsoner) [inline scipy/optimize/tests/test_slsqp.py:605]: More explicit and specific
```suggestion
        with pytest.warns(RuntimeWarning, match='were outside bounds'):
```

**Review by larsoner — APPROVED** [review]
Just some minor DRY / explicit suggestions here, again +1 to merge without them if necessary

File: scipy/optimize/tests/test_slsqp.py [hunk 5]
```diff
@@ -207,6 +208,27 @@ def test_minimize_inequality_given_vector_constraints(self):
...
         assert_allclose(res.x, [2, 1])
 
+    def test_minimize_bounded_constraint(self):
+        # when the constraint makes the solver go up against a parameter
+        # bound make sure that the numerical differentiation of the
+        # jacobian doesn't try to exceed that bound using a finite difference.
+        # gh11403
+        def c(x):
+            if x[0] < 0 or x[1] < 0 or x[0] > 1 or x[1] > 1:
+                assert False
```
[the remark this answers is not in the document]
Comment (andyfaff) [reply scipy/optimize/tests/test_slsqp.py:218]: Ack, also using this suggestion for f(x)

File: scipy/optimize/optimize.py (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py:280]: Ack.

File: scipy/optimize/tests/test_slsqp.py (hunk 4 above)
Comment (andyfaff) [reply scipy/optimize/tests/test_slsqp.py:605]: Ack

**Comment by andyfaff** [comment]
@larsoner I made the suggested changes and rebased. Everything you reviewed is now in the first commit, the changes I made are in the second.

**Comment by larsoner** [comment]
Travis is just unhappy about some linting errors that I guess we can safely ignore. Thanks @andyfaff !

**Comment by Bezerragrasi** [comment]
Any new on this issue?
I am having the follow warming when I use the SLSQP minimization 

```
RuntimeWarning: Values in x were outside bounds during a minimize step, clipping to bounds
  warnings.warn("Values in x were outside bounds during a "
```

all issues I read in the matter are from older versions of scipy, I am using scipy=1.7.3 and numpy=1.21.2. How can I maintain the minimization steps within the bounds? And any idea in why is this happening?

**Pull request merged**